### PR TITLE
Update IAS definition

### PR DESCRIPTION
**Pull request by T2Fat2Fly**
Hi Boris!
I'm [2Fat2Fly on KSP forum](https://forum.kerbalspaceprogram.com/index.php?/profile/149977-2fat2fly/) (previously Fat_Bird).

First of all thanks for your amazing addon which is still the best among the autopilot plugins for KSP IMO!
I've picked up KSP again recently and now I'm comming back to try to solve [several issues mentioned several years ago](https://forum.kerbalspaceprogram.com/index.php?/topic/124417-180-1123-atmosphereautopilot-160/&do=findComment&comment=2462486). I've split the changes to several branches to better describe them one by one. This being the first of them.

# Indicated Airspeed (IAS) Definition Update

## The issue

In the previous code, the "IAS" was defined using true airspeed (ground speed) against atmosphere density:
https://github.com/Boris-Barboris/AtmosphereAutopilot/blob/6d21fdbedcb51bf1f75c1b3c973a5bd2f755986b/AtmosphereAutopilot/Modules/ProgradeThrustController.cs#L64
Of which is actualy [equivalent airspeed (EAS) by definition](https://en.wikipedia.org/wiki/Equivalent_airspeed).

However, as quoted from the [IAS Wikipedia page](https://en.wikipedia.org/wiki/Indicated_airspeed):

> However, at typical civilian operating speeds, the aircraft's aerodynamic structure responds to dynamic pressure alone, and the aircraft will perform the same when at the same dynamic pressure. Since it is this same dynamic pressure that drives the airspeed indicator, an aircraft will always, for example, stall at the published indicated airspeed (for the current configuration) regardless of density, altitude or true airspeed.

So I think it is necessary to tell the autopilot to stick to a certain IAS, instead of a certain EAS.

## Solution

In the Prograde Thrust Controller module, I replicated [the FARc's code](https://github.com/dkavolis/Ferram-Aerospace-Research/blob/49449b5fc833eef87f1e749ef80883bc1111d7ce/FerramAerospaceResearch/FARAeroUtil.cs#L323) to calculate IAS, as below:
https://github.com/T2Fat2Fly/AtmosphereAutopilot/blob/1a909acf1974c5f59ba366451969645d80ea1655/AtmosphereAutopilot/Modules/ProgradeThrustController.cs#L68
But since the code have nothing to do with FARc, it should work with stock aerodynamic as well.

Since the PTC module uses mps (ground speed) as setpoint for thrust control, a reverse function is also included, which involved secant method for the supersonic part. Luckily it seems not to have a significant performance hit in game.

## Result

FAR stock aircraft FAR Firehound MS used as example. It seems that a good consistency with the FARc IAS is established, both subsonically and supersonically:
pic1
pic2
pic3

### CODE REVIEW

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs [hunk 1]
```diff
@@ -40,6 +40,96 @@ public struct SpeedSetpoint
...
+            pressure /= (4 * gamma * M * M - 2 * (gamma - 1));
+            pressure = Mathf.Pow(pressure, gamma / (gamma - 1));
+            pressure *= (1 - gamma + 2 * gamma * M * M) / (gamma + 1);
+
+            return pressure;
+        }
+
+        public static float mps2Ias(float mps, Vessel v)
+        {
+            float M = (float)(mps / v.speedOfSound);
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: For every division, sqrt and other possible NaN-poisoning operations, please check that we don't fall flat.
Vacuum or some other conditions may break the new code - you should probably handle it just in case.

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs [hunk 2]
```diff
@@ -40,6 +40,96 @@ public struct SpeedSetpoint
...
+            float gamma = (float) v.mainBody.atmosphereAdiabaticIndex;
+            float pressureRatio;
+            if (M <= 1)
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: Is the transition smooth/continuous? If not, why? And it may need a filter.

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs [hunk 3]
```diff
@@ -40,6 +40,96 @@ public struct SpeedSetpoint
...
+            return velocity;
+        }
+
+        public static float rayleighMachFromPitotTubeStagPressure(float pressure, float gamma)
+        {
+            float M;
+
+            float soundPres = Mathf.Pow((gamma + 1) / 2, gamma / (gamma - 1)); ;
+
+            if (pressure <= soundPres)
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: same question: is the transition smooth/continuous?

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs [hunk 4]
```diff
@@ -40,6 +40,96 @@ public struct SpeedSetpoint
...
+                do
+                {
+                    slope = (m2 - m1) / (pt2 - pt1);
+                    mnew = m2 - slope * pt2;
+                    mnew = Mathf.Max(1, mnew); // don't go subsonic
+                    m1 = m2;
+                    m2 = mnew;
+                    pt1 = pt2;
+                    pt2 = rayleighPitotTubeStagPressureSuperSonic(m2, gamma) - pressure;
+                } while (Mathf.Abs(m2 - m1) > machEpsilon);
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: Let's put a limit to iteration count and print Error to log if we reach it.
Frankly, not a fan of such heavy math in our fixedUpdate =( 
Lot's of cycles for such a petty formularity.

Does the method always converge?
Can we replace it by some precalculated mesh interpolation?

**Review by Boris-Barboris — CHANGES_REQUESTED** [review]
Ooof, indeed, heavy code.

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs (hunk 2 above)
Comment (T2Fat2Fly) [reply AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: Yes it is. At M = 1 both subsonic and supersonic pressure equals to Pow((gamma + 1) / 2, gamma / (gamma - 1)). Ferram used a simplified version of the Rayleigh formula and achieved both simpleness and continuity.

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs (hunk 4 above)
Comment (T2Fat2Fly) [reply AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: Sounds bold but I suggest using IAS alone for setpoint. Here's my 2 cents:
1. IRL aircrafts never knows their true speed against ground (and static air) hence IAS is their only choise to read & hold (along with mach at higher altitude, which is also calculated from IAS);
2. As said before, IAS plays an important role in aerodynamic and maneuverability;
3. All other speed modes in AA, can be converted to IAS with O(1) algorithms.

Comment (Boris-Barboris) [reply AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: 1). I never intended to constrain AAP to what IRL aircraft knows. I used every easily-accessible information the game gave me.
2). ...in FAR. Maybe we should only use these formulas when FAR is installed? There already exists FAR-specific code in AAP, so that would be easy to code and definitely easy my mind.
3). Nah, let's keep it in kinematic speed. Simple is good.

Comment (T2Fat2Fly) [reply AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: I came with another possible solution.
I've found a way to read IAS value directly from FAR so we may ditch the calculation code.
Then, I want to make another change: we determine the airspeed differencies from inside the Prograde Thrust Controller and ditch the need to call mps() method outside the class. So we can apply thrust controls without the need of the IAS reverse functions. It will also make PTC a StateController rather than a SISOController.

If FAR is not found, we can then use the good old EAS code.

Comment (Boris-Barboris) [reply AtmosphereAutopilot/Modules/ProgradeThrustController.cs]: Sounds good! If you feel comfortable with runtime reflections (to call FAR code without depending on it), that would be the simplest solution, IMO.

File: AtmosphereAutopilot/Reflections/FARReflections.cs [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+                // Get FAR control surface type
+                FARControllableSurfaceType = ReflectionUtils.GetType(farAssembly, "FARControllableSurface");
+                if (FARControllableSurfaceType == null)
+                {
+                    throw new Exception("AtmosphereAutopilot could not bind to FAR FARControllableSurface class");
+                }
+
+                // Load ActiveVesselIAS method from FARAPI
+                var activeVesselIASMethodInfo = ReflectionUtils.GetMethodInfo(
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Reflections/FARReflections.cs:54]: Active vessel only? What happens if you fly in formation and switch to another vessel? The autopilot of the first vessel must be autonomous.

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs [hunk 6]
```diff
@@ -403,13 +628,16 @@ float solve_thrust_req(double required_thrust, float prev_throttle)
 
         #region GUI
 
-        static readonly string[] spd_str_arr = new string[] { "OFF", "ms", "kts", "M", "ias", "kias" };
+        string[] spd_str_arr = new string[] { "m/s", "kts", "Mach", "IAS", "kIAS" };
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/ProgradeThrustController.cs:631]: spd_type_names

File: AtmosphereAutopilot/Modules/StateController.cs [hunk 7]
```diff
@@ -39,6 +39,23 @@ protected StateController(Vessel cur_vessel, string module_name, int wnd_id)
 
     }
 
+    /// <summary>
+    /// Base class of sub state controllers that is called by main state controllers
+    /// </summary>
+    public abstract class SubStateController : AutopilotModule
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/StateController.cs:45]: I don't think we need another base class. StateController is fine as base for ProgradeThrust...

Comment (Boris-Barboris) [reply AtmosphereAutopilot/Modules/StateController.cs:45]: It would be trouble only if it was sealed: then it would get caught into autopilot_module_types and registered as top-level autopilot.

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs [hunk 8]
```diff
@@ -47,77 +77,171 @@ public SpeedSetpoint(SpeedType type, float value, Vessel v)
...
+
+        
+        // unit convertions within the same system.
+
+        /// <summary>
+        /// Get speed in m/s of the current airspeed reference system.
+        /// If converting between systems, use convert() instead.
+        /// </summary>
+        /// <returns>Speed in m/s</returns>
+        public float sameSystemMps()
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/ProgradeThrustController.cs:112]: currentSystemMps

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs [hunk 9]
```diff
@@ -31,6 +31,36 @@ public enum SpeedType
         KIAS
     }
 
+    public static class SpeedTypeMethods
+    {
+        // SpeedType classification
+        public static bool isKinematicSystem(this SpeedType t)
```
Comment (Boris-Barboris) [inline AtmosphereAutopilot/Modules/ProgradeThrustController.cs:37]: I think if you would go one step further and not treat this as bool, the LoC would reduce. We actually already have three unit systems (where units in the same system are linearly convertible between eachother):

kinematic (mps, kts),
instrumental (ias, kias),
sound (mach)

Old code was treating kinematic mps as lingua franca and ran all conversions through it. That is why the old code was 'switch' and not 'if' based. Can you try and separate unit system to some tangible named entity? Either another (maybe, private) enum or class hierarchy... 

Basically, my point is, current code looks a bit ugly because the pressure/kinematic condition is too prevalent in the interface and forces you to write a lot of if clauses. And you can't extend it to new unit systems easily. What if you wanted to support both IAS and EAS?

**Review by Boris-Barboris — CHANGES_REQUESTED** [review]
May cause bonus despair, but the problems are not that small to ignore.

**Comment by T2Fat2Fly** [comment]
Oops sorry I didn't know committing on my own branch will also wake this PR up. 🤣 So no document (yet) because I'm not actually ready to file changes to the public.

File: AtmosphereAutopilot/Reflections/FARReflections.cs (hunk 5 above)
Comment (T2Fat2Fly) [reply AtmosphereAutopilot/Reflections/FARReflections.cs:54]: Another method that accept Vessel as a parameter is also included in the FAR API. I'll apply and test that one.
~~THB~~ TBH I don't know AAP works on non-focused vessel. 😂
(wtf Apple autocorrections)

File: AtmosphereAutopilot/Modules/ProgradeThrustController.cs (hunk 9 above)
Comment (T2Fat2Fly) [reply AtmosphereAutopilot/Modules/ProgradeThrustController.cs:37]: > current code looks a bit ugly because the pressure/kinematic condition is too prevalent in the interface and forces you to write a lot of if clauses

Yes and I already have trouble naming functions. (e.g. "sameSystemMps" is used to avoid confusion like "currentSpeedInCurrentSystem"😂)

Actually this commit is meant to be a temporary fix and I'm still testing with it (in my actual gameplay :P). Currently it doesn't work well in IAS mode when climbing/descending (changing in dyn. pressure must be taken into account for acceleration prediction) and I'm still looking forward to another refactoring (v3 perhaps).

Please ignore my commit on the FLC branch. Changes on that branch is meaningless until we settle down a solution on this branch. (Whew great it seems that I "forgot" to commit them.)
